streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamImplValueJoinerWithKeyTest.java

```diff
@@ -0,0 +1,229 @@
+        globalKTable = builder.globalTable(globalTopic, Consumed.with(Serdes.String(), Serdes.Integer()));
+    }
+
+    @Test
+    public void shouldIncludeKeyInStreamSteamJoinResults() {
+        leftStream.join(
+                rightStream,
+                valueJoinerWithKey,
+                joinWindows,
+                streamJoined
+        ).to(outputTopic, Produced.with(Serdes.String(), Serdes.String()));
+        // Left KV A, 3, Right KV A, 5
+        runJoinTopology(builder,
+                Collections.singletonList(KeyValue.pair("A", "A:5")),
+                false,
+                rightTopic
+        );
+    }
+
+    @Test
+    public void shouldIncludeKeyInStreamLeftJoinResults() {
+        leftStream.leftJoin(
+                rightStream,
+                valueJoinerWithKey,
+                joinWindows,
+                streamJoined
+        ).to(outputTopic, Produced.with(Serdes.String(), Serdes.String()));
+        // Left KV A, 3, Right KV A, 5
+        // TTD pipes records to left stream first, then right
+        // with TTD there's no caching, so join emits immediately with "A, 3, null" then "A, 3, 5"
+        final List<KeyValue<String, String>> expectedResults = Arrays.asList(KeyValue.pair("A", "A:3"), KeyValue.pair("A", "A:5"));
+        runJoinTopology(builder,
+                expectedResults,
+                false,
+                rightTopic
+        );
+    }
+
+    @Test
+    public void shouldIncludeKeyInStreamOuterJoinResults() {
+        leftStream.outerJoin(
+                rightStream,
+                valueJoinerWithKey,
+                joinWindows,
+                streamJoined
+        ).to(outputTopic, Produced.with(Serdes.String(), Serdes.String()));
+
+        // Left KV A, 3, Right KV A, 5
+        // TTD pipes records to left stream first, then right
+        // with TTD there's no caching, so join emits immediately with "A, 3, null" then "A, 3, 5"
+        final List<KeyValue<String, String>> expectedResults = Arrays.asList(KeyValue.pair("A", "A:3"), KeyValue.pair("A", "A:5"));
+        runJoinTopology(builder,
+                expectedResults,
+                false,
+                rightTopic
+        );
+    }
+
+    @Test
+    public void shouldIncludeKeyInStreamTableJoinResults() {
+        leftStream.join(
+            ktable,
+            valueJoinerWithKey,
+            joined
+        ).to(outputTopic, Produced.with(Serdes.String(), Serdes.String()));
+        // Left KV A, 3, Table KV A, 5
+        runJoinTopology(builder,
+                Collections.singletonList(KeyValue.pair("A", "A:5")),
+                true,
+                ktableTopic
+        );
+    }
+
+    @Test
+    public void shouldIncludeKeyInStreamTableLeftJoinResults() {
+        leftStream.leftJoin(
+            ktable,
+            valueJoinerWithKey,
+            joined
+        ).to(outputTopic, Produced.with(Serdes.String(), Serdes.String()));
+        // Left KV A, 3, Table KV A, 5
```

Comment:
Similar to above: should we send one more left record with key `C`?

Reply:
Ack